flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Periodically make materialization for the delegated state backend. */
+public class PeriodicMaterializer {
+    private static final Logger LOG = LoggerFactory.getLogger(PeriodicMaterializer.class);
+
+    /**
+     * ChangelogStateBackend only supports CheckpointType.CHECKPOINT; The rest of information in
+     * CheckpointOptions is not used in Snapshotable#snapshot(). More details in FLINK-23441.
+     */
+    private static final CheckpointOptions CHECKPOINT_OPTIONS =
+            new CheckpointOptions(
+                    CheckpointType.CHECKPOINT, CheckpointStorageLocationReference.getDefault());
+
+    /** task mailbox executor, execute from Task Thread. */
+    private final MailboxExecutor mailboxExecutor;
+
+    /** Async thread pool, to complete async phase of materialization. */
+    private final ExecutorService asyncOperationsThreadPool;
+
+    /** scheduled executor, periodically trigger materialization. */
+    private final ScheduledExecutorService periodicExecutor;
+
+    private final CheckpointStreamFactory streamFactory;
+
+    private final Snapshotable keyedStateBackend;
+
+    private final StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter;
+
+    private final AsyncExceptionHandler asyncExceptionHandler;
+
+    private final int allowedNumberOfFailures;
+
+    /** Materialization failure retries. */
+    private final AtomicInteger retries;
+
+    /** Making sure only one materialization on going at a time. */
+    private final AtomicBoolean materializationOnGoing;
+
+    private long materializedId;
+
+    private MaterializedState materializedState;
+
+    public PeriodicMaterializer(
+            MailboxExecutor mailboxExecutor,
+            ExecutorService asyncOperationsThreadPool,
+            Snapshotable keyedStateBackend,
+            CheckpointStorageWorkerView checkpointStorageWorkerView,
+            StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter,
+            AsyncExceptionHandler asyncExceptionHandler,
+            MaterializedState materializedState,
+            long periodicMaterializeInitDelay,
+            long periodicMaterializeInterval,
+            int allowedNumberOfFailures,
+            boolean materializationEnabled) {
+        this.mailboxExecutor = mailboxExecutor;
+        this.asyncOperationsThreadPool = asyncOperationsThreadPool;
+        this.keyedStateBackend = keyedStateBackend;
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.asyncExceptionHandler = asyncExceptionHandler;
+        this.periodicExecutor =
+                Executors.newSingleThreadScheduledExecutor(
+                        new ExecutorThreadFactory("periodic-materialization"));
```

Comment:
nit: add `-scheduler` to name? and ideally task info (`env.getTaskInfo().getTaskNameWithSubtasks()`)

Reply:
Yes, I think add task name is a good idea!